### ISSUE TITLE
How to inject a xUnit Fixture

### ISSUE THREAD

**Issue by diegosasw** [LightBDD/LightBDD#355]
#### Description
I am struggling to find an answer by reading the documentation for xUnit v2.

I want my scenarios to reuse a server, for example a test container of a database that I want to spin up once before all the tests run, and destroy when all the tests have finished.

The way I usually achieve this is with a Fixture

```
public class TestContainerFixture
    : IAsyncLifetime
{
    public MyContainerResult MyContainerResult { get; private set; } = null!;
    
    private MyContainer? _myTestContainer;

    public async Task InitializeAsync()
    {
        _myTestContainer= new MyTestContainer();
        MyContainerResult = await _myTestContainer.Start(); // here is where the container is instantiated,  out of scope how exactly
    }

    public Task DisposeAsync() => _myTestContainer?.Stop() ?? Task.CompletedTask;
}
```

Then I would have a collection fixture
```
[CollectionDefinition("Integration")]
public class IntegrationTestCollection
    : ICollectionFixture<TestContainerFixture>
{
    // Empty, as per https://xunit.net/docs/shared-context
}
```

My tests would have a `[Collection("Integration")]` and would receive the fixture injected
```
[Collection("Integration")]
public class GivenSampleTest
{

   private TestContainerFixture _fixture;
   private ITestOutputHelper _output;

   public GivenSampleTest(TestContainerFixture fixture, ITestOutputHelper output)
   {
          _fixture = fixture;
          _output = output;
   }
   
   [Fact]
    //... Use fixture in tests
}
```

My tests would get injected the same instance of TestContainerFixture which was initiated once.

I want to have the same behavior for LightBdd but I can't find how this integrates with xUnit class or collection fixtures, and it seems that I cannot have any complex constructor for my LightBdd tests as in

```
public class SampleTests
    : FeatureFixture
{
    private readonly ITestOutputHelper _testOutputHelper;

    public SampleTests(ITestOutputHelper testOutputHelper)
        : base()
    {
        _testOutputHelper = testOutputHelper;
    }
   /...
}
```

Is this supported? Is there any simple example on how to achieve something similar?

PS: I need my LightBdd tests to run sequentially, and to reuse the same test container

**Comment by lemonlion** [comment]
This might help:

https://github.com/LightBDD/LightBDD/wiki/What-Is-New#2-improved-global-and-scenario-level-set-up-and-tear-down

**Comment by diegosasw** [comment]
Thanks. I am afraid I don't understand this part of the documentation.
> As a final note, it is possible now to access the fixture class instance from IScenarioDecorator implementations via IScenario.Fixture property.

Is there any usage explanation?

To give more context, if I now have this

```
public class GlobalDependency
    : IGlobalResourceSetUp
{
    public string Foo => "can you retrieve this value from an injected instance?";
    
    public Task SetUpAsync()
    {
        return Task.CompletedTask;
    }

    public Task TearDownAsync()
    {
        return Task.CompletedTask;
    }
}
```

I could use that to spin up a test container and to destroy it.

Now, I would have this

```
[assembly: ConfiguredLightBddScope] // the configured scope is applied on assembly
namespace SpecificationTests.Contracts;

internal class ConfiguredLightBddScopeAttribute 
    : LightBddScopeAttribute
{
    protected override void OnConfigure(LightBddConfiguration configuration)
    {
        configuration
            .ExecutionExtensionsConfiguration()
            .RegisterGlobalSetUp<GlobalDependency>()
            .RegisterGlobalSetUp("database", SetUpDatabase, TearDownDatabase)
            .RegisterGlobalTearDown("cleanup", RunCleanup);

        configuration
            .DependencyContainerConfiguration()
            .UseDefault(c =>
            {
                c.RegisterType<GlobalDependency>(InstanceScope.Single);
            });
    }

    private Task SetUpDatabase()
    {
        return Task.CompletedTask;
    }

    private Task TearDownDatabase()
    {
        return Task.CompletedTask;
    }

    private Task RunCleanup()
    {
        return Task.CompletedTask;
    }

}
```

And I can see these methods being executed. 

Now, I could create a base class for Feature Fixture
```
public abstract class SpecificationTestBase 
    : FeatureFixture, IScenarioSetUp, IScenarioTearDown, IDisposable
{
    protected SpecificationTestBase()
    {
    }
    
    public Task OnScenarioSetUp()
    {
        return Task.CompletedTask;
    }

    public Task OnScenarioTearDown()
    {
        return Task.CompletedTask;
    }

    public void Dispose()
    {
    }
}
```

For each scenario, I could create a `Mvc.TestServer` and expose test http client and so on. But, how do I access the "Global" database configuration from each test?
In other words, if my test container which I can spin up at the beginning in `GlobalDependency` has some specific connection string or port or whatever, and I need that when configuring my TestServer for each scenario, how do I access it?

To clarify further, in above example, how would you access the content of `Foo` from any scenario execution?

Thanks.

**Comment by Suremaker** [comment]
Hi @diegosasw ,

Please take a look at tutorials:
* https://github.com/LightBDD/LightBDD.Tutorials/tree/master/WebApiWithDependenciesServiceTests
* https://github.com/LightBDD/LightBDD.Tutorials/tree/master/WebApiServiceTests

Both use static server instance that is shared across tests. Hopefully it will help you with your problem.

**Comment by Suremaker** [comment]
With Lightbdd 3.x the easiest way would be to use GlobalSetup to spawn and destroy the singular instance of your server or database, but then use static instance (singleton) to access it from the actual tests.

The DI injection could also be used to provide server access to the tests, but if you expect only a single instance, such solution will be an overhead: https://github.com/LightBDD/LightBDD/wiki/DI-Containers#di-usage

**Comment by Suremaker** [comment]
> Thanks. I am afraid I don't understand this part of the documentation.

Let me answer this.
 
> > As a final note, it is possible now to access the fixture class instance from IScenarioDecorator implementations via IScenario.Fixture property.
> 
> Is there any usage explanation?

Code: [IScenario](src/LightBDD.Core/Execution/IScenario.cs)

Please see the comments in the snippet below
```c#
 public class SampleTests  //<-- IScenario.Fixture will refer to this instance
     : SpecificationTestBase
 {
     [Scenario]
     public void ShouldSucceed()
         => Runner
             .WithContext(new SampleTestContext()) // <-- IScenario.Context will refer to this instance
             .RunScenario(
                 given => given.Foo(),
                 when => when.Foo(),
                 then => then.Foo());
  }
```

`Fixture` and `Context` properties of `IScenario` gives decorator access to instance of your test class (feature fixture) and context (if you use one),

> To give more context, if I now have this
> 
> ```
> public class GlobalDependency
>     : IGlobalResourceSetUp
> {
>     public string Foo => "can you retrieve this value from an injected instance?";
>     
>     public Task SetUpAsync()
>     {
>         return Task.CompletedTask;
>     }
> 
>     public Task TearDownAsync()
>     {
>         return Task.CompletedTask;
>     }
> }
> ```
> 
> I could use that to spin up a test container and to destroy it.
> 
> Now, I would have this
> 
> ```
> [assembly: ConfiguredLightBddScope] // the configured scope is applied on assembly
> namespace SpecificationTests.Contracts;
> 
> internal class ConfiguredLightBddScopeAttribute 
>     : LightBddScopeAttribute
> {
>     protected override void OnConfigure(LightBddConfiguration configuration)
>     {
>         configuration
>             .ExecutionExtensionsConfiguration()
>             .RegisterGlobalSetUp<GlobalDependency>()
>             .RegisterGlobalSetUp("database", SetUpDatabase, TearDownDatabase)
>             .RegisterGlobalTearDown("cleanup", RunCleanup);
> 
>         configuration
>             .DependencyContainerConfiguration()
>             .UseDefault(c =>
>             {
>                 c.RegisterType<GlobalDependency>(InstanceScope.Single);
>             });
>     }
> 
>     private Task SetUpDatabase()
>     {
>         return Task.CompletedTask;
>     }
> 
>     private Task TearDownDatabase()
>     {
>         return Task.CompletedTask;
>     }
> 
>     private Task RunCleanup()
>     {
>         return Task.CompletedTask;
>     }
> 
> }
> ```
> 
> And I can see these methods being executed.
> 
> Now, I could create a base class for Feature Fixture
> 
> ```
> public abstract class SpecificationTestBase 
>     : FeatureFixture, IScenarioSetUp, IScenarioTearDown, IDisposable
> {
>     protected SpecificationTestBase()
>     {
>     }
>     
>     public Task OnScenarioSetUp()
>     {
>         return Task.CompletedTask;
>     }
> 
>     public Task OnScenarioTearDown()
>     {
>         return Task.CompletedTask;
>     }
> 
>     public void Dispose()
>     {
>     }
> }
> ```
> 
> For each scenario, I could create a `Mvc.TestServer` and expose test http client and so on. But, how do I access the "Global" database configuration from each test? In other words, if my test container which I can spin up at the beginning in `GlobalDependency` has some specific connection string or port or whatever, and I need that when configuring my TestServer for each scenario, how do I access it?

As per my previous message, the easiest way would be to refer to it via static property.
Here is an example of server global initialization: https://github.com/LightBDD/LightBDD.Tutorials/blob/56be483f9ba8b9a09e6ff1e88c3ff777467ab31b/WebApiServiceTests/CustomerApi.ServiceTests/ConfiguredLightBddScope.cs#L18

And here is an example of it's usage from the tests: https://github.com/LightBDD/LightBDD.Tutorials/blob/master/WebApiServiceTests/CustomerApi.ServiceTests/Features/Adding_customers.Steps.cs#L23

This would be the simplest way.

> 
> This is how I want to write tests.
> 
> ```
> public class SampleTests
>     : SpecificationTestBase
> {
>     [Scenario]
>     public void ShouldSucceed()
>         => Runner
>             .WithContext(new SampleTestContext())
>             .RunScenario(
>                 given => given.Foo(),
>                 when => when.Foo(),
>                 then => then.Foo());
>     
>     [Scenario]
>     public void ShouldSucceed2()
>         => Runner
>             .WithContext(new SampleTestContext())
>             .RunScenario(
>                 given => given.Foo(),
>                 when => when.Foo(),
>                 then => then.Foo());
> }
> ```
> 
> To clarify further, in above example, **how would you access the content of `Foo` from any scenario execution?**

I am not 100% sure what you mean by `from any scenario execution`.

Assuming you use `.WithContext(new SampleTestContext())` to create new context instance, then such context instance is accessible via `given.`, `when.`, `then.` lambda parameters, so `Foo()` is really a method of `SampleTestContext()`.

Actually if you do not want to use static instances to access your global server, you can leverage `DI` and contexts in the following way:

```c#

// from your ConfiguredLightBddScopeAttribute configuration
 configuration
     .DependencyContainerConfiguration()
     .UseDefault(c =>
     {
         c.RegisterType<GlobalDependency>(InstanceScope.Single);
     });

// definition of SampleTestContext class with DI injection via constructor

public class SampleTestContext
{
  private GlobalDependency _server;
  public SampleTestContext(GlobalDependency server) { _server = server; }
  public void Foo() { }
}

// context instantiation from DI
[Scenario]
public void ShouldSucceed2()
   => Runner
       .WithContext<SampleTestContext>() // uses DI to resolve
       .RunScenario(
           given => given.Foo(), //Foo will have access to server via _server field
           when => when.Foo(),
           then => then.Foo());
```

This approach is cleaner as it does not use server access via static properties, but it is more complex as it requires usage of context classes.

Documentation:
- https://github.com/LightBDD/LightBDD/wiki/Scenario-Steps-Definition#contextual-scenarios
- https://github.com/LightBDD/LightBDD/wiki/DI-Containers#using-di-containers-in-contextual-scenarios
- https://github.com/LightBDD/LightBDD/wiki/SetUp-and-TearDown#global-setup-and-teardown

The limitation of using DI in LightBDD 3.x is that there is no support for dependency injection in test class constructors, so if you do not want to use context classes but access dependencies from test class, you would need to resolve these dependencies programmatically, and you are limited to only do that when scenario is already running (not from the constructor).

```c#
public abstract class SpecificationTestBase 
     : FeatureFixture, IScenarioSetUp, IScenarioTearDown, IDisposable
 {
     protected GlobalDependency Server { get; private set; }
     protected SpecificationTestBase()
     {
       // Server is not initialized here
       // Server cannot be resolved here as class constructor is called outside scenario
     }
     
     public Task OnScenarioSetUp()
     {
         Server = StepExecution.Current.GetScenarioDependencyResolver().Resolve<GlobalDependency>();
         return Task.CompletedTask;
     }
 
     public Task OnScenarioTearDown()
     {
         return Task.CompletedTask;
     }
 
     public void Dispose()
     {
     }
 }
```

Documentation: https://github.com/LightBDD/LightBDD/wiki/DI-Containers#programmatic-access-to-di-container

> Thanks.

**Comment by diegosasw** [comment]
@Suremaker 
thanks for your replies. I like the idea of using the Context and relying on DI at the context level. That would solve the issue of accessing the singleton global instance during test execution in a clean elegant way.

But this leaves out another piece of the puzzle: How do I access this global instance (e.g: the `IGlobalResourceSetUp`) from the `IScenarioSetUp` and `IScenarioTearDown`.

I need this because I need every Scenario to configure TestServer and override specific Services override (i.e: the `WebApplicationFactory<TEntryPoint>` where I access global resources in order to override some configuration). TestServer cannot be global, as every Scenario should be able to override some services during `WebApplicationFactory` instantiation.

The Context class cannot implement `IScenarioSetUp` and `IScenarioTearDown` as far as I can see. Its methods won't be executed, so I was looking for a way to have it at the `FeatureFixture` but here I can't inject the `IGlobalResourceSetUp` as per previous issues.

I feel I'm going in circles without having the "aha" moment.

To summarize, what I want is
- Something to run before All scenarios and after All scenarios. This is solved with the `IGlobalResourceSetUp` which uses LightBdd DI container capabilitiers.
- Something to access `IGlobalResourceSetUp` before and after Scenario execution. Using Context is great, and I like that it does not require additional complexity, but the Context does not have something to run before and after Scenario where I could configure TestServer, does it?

Thanks again for your time!

PS: I created a [sample repository](https://github.com/diegosasw/cosmosdb-masstransit-lightbdd-testing) where I am configuring two test projects: one integration tests which uses standard xUnit, CollectionDefinition, Fixture, etc., and one test project where I want to have THE SAME (or similar) approach but using LightBdd framework.

**Comment by Suremaker** [comment]
Hi @diegosasw .
I believe I provided all the necessary blocks in the previous comments, but let me summarize them below.

```c#
// to run setup before ALL scenarios and tear down after ALL scenarios
public class GlobalDependency : IGlobalResourceSetUp
{
    public Task SetUpAsync() { /* global init */  }

    public Task TearDownAsync() {/* global tear down */   }
}
```

```c#
internal class ConfiguredLightBddScopeAttribute 
    : LightBddScopeAttribute
{
    protected override void OnConfigure(LightBddConfiguration configuration)
    {
        configuration
            .ExecutionExtensionsConfiguration()
            .RegisterGlobalSetUp<GlobalDependency>(); //register GlobalDependency to run on global startup / global tear down

        configuration
            .DependencyContainerConfiguration()
            .UseDefault(c =>
            {
                c.RegisterType<GlobalDependency>(InstanceScope.Single); // register single instance of GlobalDependency (to be used by startup/tear down but also in DI injections
            });
    }
}
```

```c#
// context class, in case you'd like to use context approach
public class MyContext: IDisposable //IAsyncDisposable
{
    private GlobalDependency _dependency;
    public MyContext(GlobalDependency dependency)  // DI
    { 
        _dependency = dependency; 
       // this runs before scenario
    }

    public void Dispose () { /* this runs after scenario */ } 
    public void Given_foo() { /* access via field */ } 
}
```

```c#
public class MyFeature: FeatureFixture  //feature using context
{
    [Scenario]
    public Task My_scenario() => Runner.WithContext<MyContext>().RunScenarioAsync(x => x.Given_foo());
}

public class MyFeature2: FeatureFixture, IScenarioSetUp //feature using scenario setup
{
    private GlobalDependency _dependency;

     public Task OnScenarioSetUp()
     {
         _dependency = StepExecution.Current.GetScenarioDependencyResolver().Resolve<GlobalDependency>();
         return Task.CompletedTask;
     }

    [Scenario]
    public Task My_scenario() => Runner.RunScenarioAsync(_ => _.Given_foo());

    public void Given_foo() { /* access via field */ } 
}
```

The above code shows how to access the GlobalDependency from context class (`MyFeature`) and from `OnScenarioSetUp()` (`MyFeature2`).

> Something to access IGlobalResourceSetUp before and after Scenario execution. Using Context is great, and I like that it does not require additional complexity, but the Context does not have something to run before and after Scenario where I could configure TestServer, does it?

When you specify context via `.WithContext<MyContext>()` method, it will instantiate it for each scenario, so your "before scenario hook" is the context constructor while "after scenario hook" will be `Dispose()` / `DisposeAsync()`, assuming you add `IDisposable / IAsyncDisposable` to `MyContext`.

**Comment by Suremaker** [comment]
Here is one more example, which used two dependencies:
- `GlobalDependency` registered as singleton, started before all scenarios and disposed after,
- `LocalDependency` registered as instance per scenario, where each scenario will get its own.

The example below uses custom context (`MyContext`) in both cases.
`My_scenario_with_parameter()` shows example on how to customize LocalDependency (via MyContext) with scenario parameters.

```c#
internal class ConfiguredLightBddScopeAttribute : LightBddScopeAttribute
{
    protected override void OnConfigure(LightBddConfiguration configuration)
    {
        configuration
            .ExecutionExtensionsConfiguration()
            .RegisterGlobalSetUp<GlobalDependency>();

        configuration
            .DependencyContainerConfiguration()
            .UseDefault(c =>
            {
                c.RegisterType<GlobalDependency>(InstanceScope.Single);
                c.RegisterType<LocalDependency>(InstanceScope.Scenario);
            });
    }

}

internal class GlobalDependency : IGlobalResourceSetUp
{
    public string GlobalId { get; private set; }

    public Task SetUpAsync()
    {
        GlobalId = $"Runtime-{Random.Shared.Next(100)}";
        return Task.CompletedTask;
    }

    public Task TearDownAsync()
    {
        return Task.CompletedTask;
    }
}

internal class LocalDependency : IDisposable
{
    private readonly GlobalDependency _global;
    public string LocalId { get; private set; }
    public LocalDependency(GlobalDependency global)
    {
        _global = global;
    }

    public void Customize(string value)
    {
        LocalId = $"Local-{value} ({_global.GlobalId})";
    }

    public void Dispose()
    {
    }
}

internal class MyContext
{
    private readonly LocalDependency _local;

    public MyContext(LocalDependency local)
    {
        _local = local;
        _local.Customize("default");
    }

    public void Customize(string value) => _local.Customize(value);

    public Task Given_foo()
    {
        StepExecution.Current.Comment($"My local dependency: {_local.LocalId}");
        return Task.CompletedTask;
    }
}

public class My_feature : FeatureFixture
{
    [Scenario]
    public Task My_scenario() => Runner
        .WithContext<MyContext>()
        .RunScenarioAsync(x => x.Given_foo());

    [Scenario]
    [InlineData("A")]
    [InlineData("B")]
    public Task My_scenario_with_parameter(string parameter) => Runner
        .WithContext<MyContext>(c => c.Customize(parameter))
        .RunScenarioAsync(x => x.Given_foo());
}
```

Sample output:
```
Summary:
	Test execution start time       : 2024-11-08 00:04:15 UTC
	Test execution end time         : 2024-11-08 00:04:15 UTC
	Test execution time             : 24ms
	Test execution time (aggregated): 72ms
	Number of features              : 1
	Number of scenarios             : 3
	Passed scenarios                : 3
	Bypassed scenarios              : 0
	Failed scenarios                : 0
	Ignored scenarios               : 0
	Number of steps                 : 3
	Passed steps                    : 3
	Bypassed steps                  : 0
	Failed steps                    : 0
	Ignored steps                   : 0
	Not Run steps                   : 0

Feature: My feature

	Scenario: My scenario - Passed (24ms)
		Step 1: GIVEN foo - Passed (6ms)
		Comments:
			Step 1: My local dependency: Local-default (Runtime-6)

	Scenario: My scenario with parameter "A" - Passed (24ms)
		Step 1: GIVEN foo - Passed (6ms)
		Comments:
			Step 1: My local dependency: Local-A (Runtime-6)

	Scenario: My scenario with parameter "B" - Passed (24ms)
		Step 1: GIVEN foo - Passed (6ms)
		Comments:
			Step 1: My local dependency: Local-B (Runtime-6)
```

**Comment by Suremaker** [comment]
Closing as I have not received any more comments.

**Comment by macco3k** [comment]
Is there any way to add singleton dependencies via LightBDD's DI support who would in turn depend on the global dependency without re-creating them in the context ctor?

For example, say I'm using test containers for the db and I need to start the container up as a global set up. Then I have repository classes depending on this db. Can I register them all as singletons somehow without having to re-create them in the context ctor? The tricky part here is that the db must be initialized for the repositories to get a connection string to connect to.

**Comment by Suremaker** [comment]
Hi @macco3k , yes it is. All DI implementations supports singletons. Please see documentation: https://github.com/LightBDD/LightBDD/wiki/DI-Containers#di-implementations
